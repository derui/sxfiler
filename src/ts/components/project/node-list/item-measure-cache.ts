export type Measure = {
  width: number;
  height: number;
};

export class ItemMeasureCache {
  private cache: Map<number, Measure> = new Map();

  public set(key: number, e: HTMLElement | null) {
    if (!e) {
      return;
    }

    const measure = {
      width: e.clientWidth,
      height: e.clientHeight,
    };

    this.cache.set(key, measure);
  }

  public get(key: number) {
    return this.cache.get(key);
  }

  public clear() {
    this.cache.clear();
  }
}
