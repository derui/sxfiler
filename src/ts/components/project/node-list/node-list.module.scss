@import "../../ui/commom-style/_colors.scss";
@import "../../ui/commom-style/_variables.scss";
@import "../../ui/commom-style/reset.module.css";

.root {
  display: grid;
  grid-template-rows: calc( #{$baseSpace} * 2 + 1rem ) 1fr;

  width: 100%;
  height: 100%;
  margin: 0;
  padding: 0;

  color: $sol_base1;
  overflow: hidden;
  background-color: $sol_base03;

  font-size: 1rem;
}

.resizeContainer {
  width: 100%;
  height: 100%;
  margin: 0;
  padding: 0;
}

.header {
  flex: 0 0 auto;
  padding: $baseSpace 0;
  margin: 0;

  font-weight: bold;
  color: $sol_base1;
  background-color: $sol_base03;
  font-family: monospace;

  border-bottom: 1px solid $sol_base01;

  &[data-focused="true"] {
    color: $sol_yellow;
  }

}

.list {
  composes: list-base from "../../ui/list/list.module.scss";

  font-size: 1rem;
  border-right: 1px solid $sol_base0;

  &:last-of-type {
    border-right: none;
  }
}

.empty {
  flex: 1 1 auto;
  margin: auto 0;
  text-align: center;

  color: $sol_red;

  &:after {
    content: "no any node";
    font-family: monospace;
  }
}
