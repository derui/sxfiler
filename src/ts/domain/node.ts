// define type and operations for node
import { FileStat } from "./file-stat";

export interface Node {
  id: string;
  name: string;
  stat: FileStat;
  parentDirectory: string;
  linkPath?: string;
}
