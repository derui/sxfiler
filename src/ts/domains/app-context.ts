import { UIContext } from "@/types/ui-context";

export type AppContext = {
  readonly current: UIContext;
  readonly subContexts: UIContext[];
};

/**
   create AppContext
 */
export function createAppContext({
  current,
  subContexts,
}: {
  current: UIContext;
  subContexts?: UIContext[];
}): AppContext {
  return {
    current,
    subContexts: Array.from(subContexts || []),
  };
}

/**
   change context
 */
export function changeCurrent(context: UIContext) {
  return (state: AppContext) =>
    createAppContext({
      current: context,
      subContexts: state.subContexts,
    });
}

/**
   Add a context to AppContext
 */
export function addSubContext(context: UIContext) {
  return (state: AppContext) => {
    const tmpSet = new Set(state.subContexts);
    tmpSet.add(context);

    return createAppContext({
      current: state.current,
      subContexts: Array.from(tmpSet.values()),
    });
  };
}

/**
   remove the context from AppContext
 */
export function removeSubContext(context: UIContext) {
  return (state: AppContext) => {
    const tmpSet = new Set(state.subContexts);
    tmpSet.delete(context);

    return createAppContext({
      current: state.current,
      subContexts: Array.from(tmpSet.values()),
    });
  };
}
