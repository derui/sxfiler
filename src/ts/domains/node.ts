// define type and operations for node
import { FileStat } from "./file-stat";

interface FactoryArg {
  id: string;
  name: string;
  stat: FileStat;
  parentDirectory: string;
  linkPath?: string;
}

/**
 * create node
 * @param args
 */
export function create(args: FactoryArg) {
  return new NodeImpl(args.id, args.name, args.stat, args.parentDirectory, args.linkPath);
}

export interface Node {
  readonly id: string;
  readonly name: string;
  readonly stat: FileStat;
  readonly parentDirectory: string;
  readonly linkPath?: string;
}

class NodeImpl implements Node {
  /**
   * create new Node
   * @param id id of node
   * @param name file or directory name of node
   * @param stat stats of node
   * @param parentDirectory parent directory of node
   * @param linkPath link target path if node is link
   */
  constructor(
    public readonly id: string,
    public readonly name: string,
    public readonly stat: FileStat,
    public readonly parentDirectory: string,
    public readonly linkPath?: string
  ) {}
}
