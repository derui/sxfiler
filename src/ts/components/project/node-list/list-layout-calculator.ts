import { ItemMeasureCache, Measure } from "./item-measure-cache";
import { unwrap } from "../../../utils";

export type ItemLayout = {
  index: number;
  height: number;
  offset: number;
};

type ItemMetadata = {
  offset: number;
  size: Measure;
};

export type Layout = {
  // total height of layout
  totalHeight: number;
  startIndex: number;
  stopIndex: number;
  itemLayouts: Array<ItemLayout>;
};

type InstanceProps = {
  itemMetadataMap: Map<number, ItemMetadata>;
  lastMeasuredIndex: number;
  estimatedItemSize: number;
};

// get and cache item metadata.
// This function do not handle item that have not measured, so the size of it is estimated.
function getItemMetadata(
  targetIndex: number,
  instanceProps: InstanceProps,
  getItemSize: (index: number) => Measure | undefined
): ItemMetadata {
  const { lastMeasuredIndex, itemMetadataMap } = instanceProps;
  if (targetIndex < 0) {
    throw Error(`index must be greater equal than 0: ${targetIndex}`);
  }

  if (targetIndex > lastMeasuredIndex) {
    let lastMetadata =
      lastMeasuredIndex < 0
        ? {
            offset: 0,
            size: {
              width: 0,
              height: 0,
            },
          }
        : unwrap(itemMetadataMap.get(lastMeasuredIndex));
    let offset = lastMetadata.offset + lastMetadata.size.height;

    for (let index = lastMeasuredIndex + 1; index <= targetIndex; index++) {
      let size = getItemSize(index);

      // If item have not measured, can not count last measured index
      if (!size) {
        itemMetadataMap.set(index, {
          offset,
          size: {
            width: instanceProps.estimatedItemSize,
            height: instanceProps.estimatedItemSize,
          },
        });
        offset = offset + instanceProps.estimatedItemSize;
      } else {
        itemMetadataMap.set(index, { offset, size });
        offset = offset + size.height;

        instanceProps.lastMeasuredIndex =
          instanceProps.lastMeasuredIndex === index - 1 ? index : instanceProps.lastMeasuredIndex;
      }
    }
  }
  return unwrap(itemMetadataMap.get(targetIndex));
}

// find index of the item having nearest offset given it.
function findStartIndexForOffset(offset: number, instanceProps: InstanceProps): number {
  const { lastMeasuredIndex, itemMetadataMap } = instanceProps;

  let high = lastMeasuredIndex;
  let low = 0;
  let searchIndex = 0;
  while (low < high) {
    searchIndex = Math.floor((high - low) / 2) + low;
    let itemOffset = unwrap(itemMetadataMap.get(searchIndex)).offset;

    if (itemOffset < offset) {
      low = searchIndex;
    } else if (itemOffset > offset) {
      high = searchIndex;
    } else {
      break;
    }
  }

  return searchIndex;
}

function findStopIndexForStartIndex(windowHeight: number, startIndex: number, instanceProps: InstanceProps): number {
  let height = 0;
  let index = startIndex;

  while (height < windowHeight) {
    let metadata = instanceProps.itemMetadataMap.get(index);

    if (metadata) {
      height += metadata.size.height;
    } else {
      height += instanceProps.estimatedItemSize;
    }

    index++;
  }

  return index;
}

type CalculationArguments = {
  cache: ItemMeasureCache;
  currentCursorIndex: number;
  windowHeight: number;
  currentLayout?: Layout;
};

export class ListLayoutCalculator {
  private instanceProps: InstanceProps = {
    itemMetadataMap: new Map(),
    lastMeasuredIndex: -1,
    estimatedItemSize: 50,
  };

  constructor(args: { estimatedItemSize?: number } = {}) {
    this.instanceProps.estimatedItemSize = args.estimatedItemSize ? args.estimatedItemSize : 50;
  }

  private getItemSize(cache: ItemMeasureCache, index: number) {
    const size = cache.get(index);

    if (size) {
      return size;
    }
  }

  // get offset based of current cursor, and preload metadatas to fill list window
  private calculateOffsetFromCurrentCursor(cache: ItemMeasureCache, windowHeight: number, currentCursorIndex: number) {
    const getItemSize = (index: number) => this.getItemSize(cache, index);
    let metadataForCurrentCursor = getItemMetadata(currentCursorIndex, this.instanceProps, getItemSize);
    let offset = metadataForCurrentCursor.offset;

    // load item size filling window if it exists
    let height = 0;
    let index = currentCursorIndex;
    while (height < windowHeight) {
      let metadata = getItemMetadata(index, this.instanceProps, getItemSize);
      height += metadata.size.height;
      index++;
    }

    return offset;
  }

  // get item layouts in list
  private resolveItemLayouts(startIndex: number, stopIndex: number): Array<ItemLayout> {
    const ret = [];
    for (let index = startIndex; index < stopIndex; index++) {
      let metadata = this.instanceProps.itemMetadataMap.get(index);

      if (metadata) {
        ret.push({
          index,
          height: metadata.size.height,
          offset: metadata.offset,
        });
      } else if (ret.length > 0) {
        ret.push({
          index,
          height: this.instanceProps.estimatedItemSize,
          offset: 0,
        });
      }
    }

    return ret;
  }

  // calculate layout of the list
  calculateLayout({ cache, currentCursorIndex, windowHeight, currentLayout }: CalculationArguments): Layout {
    // when current layout that calculated before passed, and index of the cursor contains
    // start/stop, no need calculation
    if (
      currentLayout &&
      currentLayout.startIndex <= currentCursorIndex &&
      currentCursorIndex < currentLayout.stopIndex
    ) {
      return { ...currentLayout };
    }

    const targetOffset = this.calculateOffsetFromCurrentCursor(cache, windowHeight, currentCursorIndex);
    const startIndex = findStartIndexForOffset(targetOffset, this.instanceProps);
    const stopIndex = findStopIndexForStartIndex(windowHeight, startIndex, this.instanceProps);
    const itemLayouts = this.resolveItemLayouts(startIndex, stopIndex);
    const totalHeight = itemLayouts.reduce((accum, itemLayout) => accum + itemLayout.height, 0);

    return {
      totalHeight,
      startIndex,
      stopIndex,
      itemLayouts,
    };
  }

  // reset cached
  reset() {
    this.instanceProps = { ...this.instanceProps, itemMetadataMap: new Map(), lastMeasuredIndex: -1 };
  }
}
