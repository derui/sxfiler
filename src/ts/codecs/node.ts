import { NodeObject, createNode } from "../domains/node";
import { encode as encodeFileStat, TypeOnRPC as FileStatOnRPC } from "./file-stat";

// define codec that is between filer domain and RPC

export type NodeOnRPC = {
  id: string;
  parent: string;
  name: string;
  stat: FileStatOnRPC;
  linkPath?: string;
};

/**
   encode node object from RPC to frontend domain.

   @param obj JSON representation for node
   @return Node object
 */
export const encode = (obj: NodeOnRPC & { marked: boolean }): NodeObject => {
  const stat = encodeFileStat(obj.stat);

  return createNode({ ...obj, stat, parentDirectory: obj.parent });
};
