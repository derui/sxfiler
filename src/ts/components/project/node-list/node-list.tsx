import * as React from "react";

import { Node } from "../../../domains/node";
import * as Element from "../../ui/element/element";
import * as List from "../../ui/list/list";
import { Component as NodeItem } from "../node-item/node-item";
import { Component as RootRef } from "../../ui/root-ref/root-ref";
import AutoSizer from "../../../libs/auto-sizer";

import { ItemMeasureCache } from "./item-measure-cache";
import { ListLayoutCalculator, Layout } from "./list-layout-calculator";

// eslint-disable-next-line @typescript-eslint/no-var-requires
const styles: ClassNames = require("./node-list.module.scss");

export interface ClassNames {
  root?: string;
  resizeContainer?: string;
  header?: string;
  list?: string;
  empty?: string;
}

interface HeaderProps {
  className?: string;
  directory: string;
  focused: boolean;
}

const HeaderElement = Element.createComponent({ tagName: "header" });

/**
 * component definition for header of file list
 */
const Header: React.FunctionComponent<HeaderProps> = ({ className, directory, focused }) => {
  return (
    <HeaderElement className={className} data-focused={focused}>
      {directory}
    </HeaderElement>
  );
};

export interface Props {
  location: string;
  nodes: Node[];
  cursor: number;
  focused: boolean;
}

export type ElementType = React.ReactElement<Props, React.FunctionComponent<Props>>;

type State = {
  needRender: boolean;
};

export class Component extends React.PureComponent<Props, State> {
  private itemMeasureCache = new ItemMeasureCache();
  private layoutCalculator: ListLayoutCalculator = new ListLayoutCalculator({
    estimatedItemSize: 24,
  });
  private currentLayout: Layout | undefined = undefined;

  state = { needRender: false };

  public componentDidMount() {
    this.setState({ needRender: true });
  }

  private makeListItems(height: number): JSX.Element[] {
    const { nodes, cursor, focused } = this.props;
    const layout = this.layoutCalculator.calculateLayout({
      cache: this.itemMeasureCache,
      currentCursorIndex: cursor,
      windowHeight: height,
      currentLayout: this.currentLayout,
    });

    return layout.itemLayouts.map(({ index }) => {
      const selected = cursor === index && focused;
      return (
        <RootRef rootRef={e => this.itemMeasureCache.set(index, e)}>
          <NodeItem key={index} item={nodes[index]} selected={selected} />
        </RootRef>
      );
    });
  }

  public render() {
    const { nodes, focused } = this.props;

    return (
      <AutoSizer className={styles.resizeContainer}>
        {({ height }) => {
          return (
            <Element.Component className={styles.root}>
              <Header key="header" className={styles.header} directory={this.props.location} focused={focused} />
              {nodes.length === 0 ? (
                <div className={styles.empty} />
              ) : (
                <List.Component className={styles.list} key="body">
                  {this.makeListItems(height)}
                </List.Component>
              )}
            </Element.Component>
          );
        }}
      </AutoSizer>
    );
  }
}
