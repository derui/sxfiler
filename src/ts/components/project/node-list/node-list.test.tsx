import * as React from "react";
import renderer from "react-test-renderer";

import { Component as T } from "./node-list";

import { create } from "../../../domains/node";

import FileStatFactory from "../../../domains/file-stat-factory";

function makeNode(name: string, isDirectory = false, isSymlink = false) {
  return create({
    id: "node",
    name,
    marked: false,
    stat: FileStatFactory.create({
      mode: "644",
      uid: 1000,
      gid: 1000,
      atime: "0",
      ctime: "0",
      mtime: "0",
      size: "10",
      isDirectory: isDirectory,
      isFile: !isDirectory && !isSymlink,
      isSymlink: isSymlink,
    }),
    parentDirectory: "/",
  });
}

describe("Project", () => {
  describe("Node List", () => {
    it("should not print before resized", () => {
      const nodes = [makeNode("file.txt")];
      const tree = renderer.create(<T nodes={nodes} location="loc" cursor={0} focused={false} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("should select a node locating same the index of cursor when focused", () => {
      const nodes = [makeNode("file.txt")];
      const tree = renderer.create(<T nodes={nodes} location="loc" cursor={0} focused={true} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("should show dummy content when nodes is empty", () => {
      const tree = renderer.create(<T nodes={[]} location="loc" cursor={0} focused={true} />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
