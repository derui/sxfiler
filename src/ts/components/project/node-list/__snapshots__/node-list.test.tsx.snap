// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Node List should not print before resized 1`] = `
<div
  className="resizeContainer"
  style={
    Object {
      "overflow": "visible",
    }
  }
/>
`;

exports[`Project Node List should select a node locating same the index of cursor when focused 1`] = `
<div
  className="resizeContainer"
  style={
    Object {
      "overflow": "visible",
    }
  }
/>
`;

exports[`Project Node List should show dummy content when nodes is empty 1`] = `
<div
  className="resizeContainer"
  style={
    Object {
      "overflow": "visible",
    }
  }
/>
`;
